lib/rewrite/rewrite-set-tag.c

```diff
@@ -35,45 +38,106 @@ struct _LogRewriteSetTag
   LogRewrite super;
   LogTagId tag_id;
   gboolean value;
+  LogTemplate *tag_template;
+  LogTemplateOptions template_options;
 };
 
+LogTemplateOptions *
+log_rewrite_set_tag_get_template_options(LogRewrite *s)
+{
+  LogRewriteSetTag *self = (LogRewriteSetTag *) s;
+  return &self->template_options;
+}
+
+static gboolean
+_init(LogPipe *s)
+{
+  LogRewriteSetTag *self = (LogRewriteSetTag *) s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (!log_rewrite_init_method(s))
+    return FALSE;
+
+  log_template_options_init(&self->template_options, cfg);
+  return TRUE;
+}
+
 static void
-log_rewrite_set_tag_process(LogRewrite *s, LogMessage **pmsg, const LogPathOptions *path_options)
+_set_tag(LogRewriteSetTag *self, LogMessage *msg)
+{
+  if (self->tag_id != LOG_TAGS_UNDEF)
+    {
+      log_msg_set_tag_by_id_onoff(msg, self->tag_id, self->value);
+      return;
+    }
+
+  GString *result = scratch_buffers_alloc();
+
+  LogTemplateEvalOptions options = {&self->template_options, LTZ_SEND, 0, NULL, LM_VT_STRING};
+  log_template_format(self->tag_template, msg, &options, result);
+
+  LogTagId tag_id = log_tags_get_by_name(result->str);
+  log_msg_set_tag_by_id_onoff(msg, tag_id, self->value);
+}
+
+static void
+_process(LogRewrite *s, LogMessage **pmsg, const LogPathOptions *path_options)
 {
   LogRewriteSetTag *self = (LogRewriteSetTag *) s;
 
   log_msg_make_writable(pmsg, path_options);
-  log_msg_set_tag_by_id_onoff(*pmsg, self->tag_id, self->value);
+
+  _set_tag(self, *pmsg);
 }
 
 static LogPipe *
-log_rewrite_set_tag_clone(LogPipe *s)
+_clone(LogPipe *s)
 {
   LogRewriteSetTag *self = (LogRewriteSetTag *) s;
   LogRewriteSetTag *cloned;
 
-  cloned = g_new0(LogRewriteSetTag, 1);
-  log_rewrite_init_instance(&cloned->super, s->cfg);
-  cloned->super.super.clone = log_rewrite_set_tag_clone;
-  cloned->super.process = log_rewrite_set_tag_process;
+  cloned = (LogRewriteSetTag *) log_rewrite_set_tag_new(self->tag_template, self->value, self->super.super.cfg);
 
   if (self->super.condition)
     cloned->super.condition = filter_expr_clone(self->super.condition);
 
-  cloned->tag_id = self->tag_id;
-  cloned->value = self->value;
   return &cloned->super.super;
 }
 
+static void
+_free(LogPipe *s)
+{
+  LogRewriteSetTag *self = (LogRewriteSetTag *) s;
+
+  log_template_unref(self->tag_template);
```

Comment:
This `unref()` is does not have its `ref()` counterpart in `new()` if `log_template_is_literal_string(tag_template)` is true.

I had another comment in the grammar file about a `ref()`-`unref()` inconsistency. I am starting to get a feeling that we should consistently `ref()` the template in `new()`, even if we will not use it later.

Reply:
If tag_template is literal, we won't even save the template, it will be null, which unref will ignore.

Reply:
This is also resolved, as I've removed template_options from set-tag().